### PR TITLE
A complete refactor of the plotting code

### PR DESCRIPTION
**Pull request by levmckinney**
The plotting code is now broken down into a set of classes:

* A `PredictionTrajectory` which represents the set of distributions by layer within the model. This class include multiple methods for visualizing different aspects of the trajectory.
* A `TrajectoryStatistic` which represents a statistic derived from the prediction trajectory that can be plotted.
* `TrajectroyLabels` a set of labels on top of a statistic to aid in visualizations.

This refactors also adds a notebook that demonstrates the new functionality `prediction_trajectory.ipynb` in the tutorials section of the docs.

### CODE REVIEW

**Comment by UlisseMini** [comment]
Haven't read everything super deeply and I'm not familiar with the codebase, but this looks syntactically wrong:
```python
        js_div = 0.5 * np.sum(
            self.probs * (self.log_probs - other.log_probs), axis=-1
        ) + 0.5 * np.sum(self.probs * (self.log_probs - self.log_probs), axis=-1) # always zero? typo?
```

Other then that looks good! I like this refactor. Will make using `HookedTransformer` from transformerlens much easier :-)

**Comment by levmckinney** [comment]
@UlisseMini thank you and good catch!

**Comment by UlisseMini** [comment]
It would probably be good to have a `from_lens_and_hidden` method. It seems the only time the model is used in `from_lens_and_model` is in the first three lines. Main workarounds (for people who's models don't support `output_hidden_states`) would be adding it as a method or constructing a `PredictionTrajectory` manually.

**Comment by levmckinney** [comment]
> It would probably be good to have a `from_lens_and_hidden` method. It seems the only time the model is used in `from_lens_and_model` is in the first three lines. Main workarounds (for people who's models don't support `output_hidden_states`) would be adding it as a method or constructing a `PredictionTrajectory` manually. Thoughts?

This seems like a good feature. Can you open a PR adding this? You know your requirements better than I do. I'd be happy to review it. I'm hopping to get this PR merged today.

File: tests/plotting/test_prediction_trajectory.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+@pytest.fixture
+def lens(model_and_tokenizer):
+    model, _ = model_and_tokenizer
+    return LogitLens(model)
+
+
+def test_prediction_trajectory_from_lens_and_model_smoke(model_and_tokenizer, lens):
+    _, tokenizer = model_and_tokenizer
+    input_ids = tokenizer.encode("Hello world!")
+    model, tokenizer = model_and_tokenizer
```
Comment (taufeeque9) [inline tests/plotting/test_prediction_trajectory.py:64]: Why not get `model` on line 62 as well?

**Review by taufeeque9 — APPROVED** [review]
Looks good! Left a few minor comments.

**Pull request merged**